packages/block-editor/src/components/block-context/index.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext } from '@wordpress/element';
+
+/** @typedef {import('react').ReactNode} ReactNode */
+
+/**
+ * @typedef BlockContextProviderProps
+ *
+ * @property {Record<string,*>} value    Context value to merge with current
+ *                                       value.
+ * @property {ReactNode}        children Component children.
+ */
+
+/** @type {import('react').Context<Record<string,*>>} */
+const Context = createContext( {} );
+
+/**
+ * Component which merges passed value with current consumed block context.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-context/README.md
+ *
+ * @param {BlockContextProviderProps} props
+ */
+export function BlockContextProvider( { value, children } ) {
+	const context = useContext( Context );
+
+	return (
+		<Context.Provider
+			value={ { ...context, ...value } }
```

Comment:
The provider would always receive a new value on every render, and re-render consumers. Any performance concerns around that? 

It could be an option to `useMemo` to compute the new context value.

Reply:
> It could be an option to `useMemo` to compute the new context value.

Memoized as of d97e93612.